[WebXR] Catch undeleted GL objects with OwnedGLObject

djg opened the pull request:
#### 0a69bc3f00f0de42638cd745f3ccd1254c5014cc
<pre>
[WebXR] Catch undeleted GL objects with OwnedGLObject
<a href="https://bugs.webkit.org/show_bug.cgi?id=246841">https://bugs.webkit.org/show_bug.cgi?id=246841</a>
rdar://101410305

Reviewed by Dean Jackson.

Use OwnedGLObject to ensure that OpenGL objects are correctly released
to avoid leaks. OwnedGLObject provides a reset/release interface similar
to std::unique_ptr. The results of reset() and release() are marked
[[nodiscard]] to signal the old value must be correctly handled. In
addition, release() must be called to disarm OwnedGLObject destructor
ASSERT.

* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
(WebCore::WebXROpaqueFramebuffer::~WebXROpaqueFramebuffer):
(WebCore::WebXROpaqueFramebuffer::startFrame):
(WebCore::WebXROpaqueFramebuffer::setupFramebuffer):
* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.h:
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
(WebCore::OwnedGLObject::OwnedGLObject):
(WebCore::OwnedGLObject::~OwnedGLObject):
(WebCore::OwnedGLObject::operator=):
(WebCore::OwnedGLObject::operator bool const):
(WebCore::OwnedGLObject::operator PlatformGLObject const):
(WebCore::OwnedGLObject::deleteObject):
(WebCore::OwnedGLObject::makeObject):

Canonical link: <a href="https://commits.webkit.org/255900@main">https://commits.webkit.org/255900@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/40e626c3c98529d2e0af5dc2c9704d4f65f06b32

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93820 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3014 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/24397 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103455 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/163781 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/97813 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3030 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31254 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86143 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/99464 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99485 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/2164 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80248 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/29183 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/84090 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/83793 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72137 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37648 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/17621 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35509 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/18881 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4071 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39388 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/41454 "Found 1 new test failure: imported/w3c/web-platform-tests/workers/interfaces/SharedWorkerGlobalScope/name/setting.html (failure)") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41322 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38112 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

grorg commented on Source/WebCore/platform/graphics/GraphicsTypesGL.h:
```diff
@@ -68,6 +68,38 @@ using GCGLContext = void*;
 typedef unsigned GLuint;
 #endif
 
+struct OwnedGLObject {
```
I wonder if this would be more user-friendly if they were specific Owners for each type of GL object. Then they could call the respective `deleteFoo` method themselves.

djg replied:
I'm glad you asked! I have a patch that does that. I wasn't exactly happy with it and started with this intermediate step.

> How did this get created without a bugzilla and radar? My guess: you started the work, already had a branch, and `git webkit` then skipped asking you to create one?

Exactly this. I added the bug ids and pushed but that doesn't appear to update the initial comment.

grorg commented on Source/WebCore/platform/graphics/GraphicsTypesGL.h:
```diff
@@ -68,6 +68,38 @@ using GCGLContext = void*;
+    OwnedGLObject(OwnedGLObject&&) = delete;
+    ~OwnedGLObject()
+    {
+        ASSERT(!m_object, "OpenGL object will leak");
```
"may" leak?

grorg replied:
Or maybe "Have you explicitly deleted this object? If so, call reset(0)."

grorg commented on Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
```diff
@@ -71,16 +71,16 @@ WebXROpaqueFramebuffer::~WebXROpaqueFramebuffer()
 #if USE(IOSURFACE_FOR_XR_LAYER_DATA)
         if (m_opaqueTexture)
-            gl->deleteTexture(m_opaqueTexture);
+            gl->deleteTexture(m_opaqueTexture.release());
```
I'm trying to work out if it would be better to have the m_opaqueTexture destructor call deleteTexture. But then it would need to hold on to a reference to the GLContext, which is sucky.

djg replied:
I had a version that was `m_opaqueTexture.deleteObject(gl)` so exact handle didn't have to carry a pointer-sized reference to `GraphicsContextGL` around.

djg committed:
[WebXR] Catch undeleted GL objects with OwnedGLObject
https://bugs.webkit.org/show_bug.cgi?id=246841
rdar://101410305

Reviewed by Dean Jackson.

Use OwnedGLObject to ensure that OpenGL objects are correctly released
to avoid leaks. OwnedGLObject provides a reset/release interface similar
to std::unique_ptr. The results of reset() and release() are marked
[[nodiscard]] to signal the old value must be correctly handled. In
addition, release() must be called to disarm OwnedGLObject destructor
ASSERT.

* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
(WebCore::WebXROpaqueFramebuffer::~WebXROpaqueFramebuffer):
(WebCore::WebXROpaqueFramebuffer::startFrame):
(WebCore::WebXROpaqueFramebuffer::setupFramebuffer):
* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.h:
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
(WebCore::OwnedGLObject::OwnedGLObject):
(WebCore::OwnedGLObject::~OwnedGLObject):
(WebCore::OwnedGLObject::operator=):
(WebCore::OwnedGLObject::operator bool const):
(WebCore::OwnedGLObject::operator PlatformGLObject const):
(WebCore::OwnedGLObject::deleteObject):
(WebCore::OwnedGLObject::makeObject):

Canonical link: https://commits.webkit.org/255900@main
